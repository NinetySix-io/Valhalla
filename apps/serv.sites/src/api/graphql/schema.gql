# ------------------------------------------------------
# THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
# ------------------------------------------------------

directive @key(fields: String!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE

directive @extends on OBJECT | INTERFACE

directive @external on OBJECT | FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @provides(fields: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

directive @link(url: String!, import: [link__Import]) on SCHEMA

directive @tag(name: String!) repeatable on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

directive @override(from: String!) on FIELD_DEFINITION

"""Use to group multiple elements together"""
type ComponentSchema {
  """Date entity was created"""
  createdAt: DateTime!

  """Date entity was updated"""
  updatedAt: DateTime!

  """Identifier of the entity"""
  id: ID!

  """Edit status"""
  status: EditStatus!

  """Account ID"""
  createdBy: String!

  """Account ID"""
  updatedBy: String!

  """Component's name"""
  name: String!

  """Thumbnail URI"""
  thumbnailUrl: String
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

enum EditStatus {
  ACTIVE
  DRAFT
  ARCHIVED
}

type PageSchema {
  """Date entity was created"""
  createdAt: DateTime!

  """Date entity was updated"""
  updatedAt: DateTime!

  """Identifier of the entity"""
  id: ID!

  """Page url"""
  slug: String

  """Edit status"""
  status: EditStatus!

  """Page title"""
  title: String!

  """Whether the title should be independent or part of the title template"""
  isLoneTitle: Boolean

  """Page description"""
  description: String

  """Account ID of creator"""
  createdBy: String!

  """Account ID of updater"""
  updatedBy: String!
}

"""Base html element"""
type ElementSchema {
  """Date entity was created"""
  createdAt: DateTime!

  """Date entity was updated"""
  updatedAt: DateTime!

  """Identifier of the entity"""
  id: ID!

  """When is root, parent is not an element"""
  isRoot: Boolean

  """Parent of element"""
  parent: String!
  updatedBy: String!

  """Element type"""
  type: ElementType!

  """Additional properties"""
  props: JSON
}

enum ElementType {
  meta
  style
  link
  title
  address
  article
  aside
  footer
  header
  h1
  h2
  h3
  h4
  h5
  h6
  main
  nav
  section
  blockquote
  dd
  div
  dl
  dt
  figcaption
  figure
  hr
  li
  menu
  ol
  p
  pre
  ul
  a
  abbr
  b
  bdi
  bdo
  br
  cite
  code
  data
  dfn
  em
  i
  kbd
  mark
  q
  rp
  rt
  ruby
  s
  samp
  small
  span
  strong
  sub
  sup
  time
  u
  var
  wbr
  area
  audio
  img
  map
  track
  video
  embed
  iframe
  object
  picture
  portal
  source
  svg
  math
  canvas
  noscript
  script
}

"""
The `JSON` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSON

type SiteSchema {
  """Date entity was created"""
  createdAt: DateTime!

  """Date entity was updated"""
  updatedAt: DateTime!

  """Identifier of the entity"""
  id: ID!

  """Site name"""
  name: String!

  """Site status"""
  status: SiteStatus!

  """Site url"""
  url: String

  """Account ID of creator"""
  createdBy: String!

  """Account ID of last updater"""
  updatedBy: String!

  """Site owner"""
  ownBy: String!
}

enum SiteStatus {
  DEPLOYED
  PENDING
  SUSPENDED
}

type HierarchicalElementResponse {
  """Identifier of the entity"""
  id: ID!

  """Element type"""
  type: ElementType!

  """Additional properties"""
  props: JSON
}

type PageUpdatedResponse {
  """Identifier of the entity"""
  id: ID!

  """Edit status"""
  status: EditStatus!
}

type SiteUpdatedResponse {
  """Identifier of the entity"""
  id: ID!

  """Site status"""
  status: SiteStatus!
}

type Query {
  siteList: [SiteSchema!]!
  site(id: String!): SiteSchema!
  pageList(siteId: String!): [PageSchema!]!
  page(siteId: String!, id: String!): PageSchema!
  component(id: String!): ComponentSchema!
  componentList: [ComponentSchema!]!
  flatElementList(filter: PageContextInput!): [ElementSchema!]!
  hierarchicalElementList(filter: PageContextInput!): [HierarchicalElementResponse!]!
}

input PageContextInput {
  """Site ID"""
  siteId: String!

  """Page ID"""
  pageId: String!
}

type Mutation {
  createSite(input: CreateSiteInput!): SiteUpdatedResponse!
  updateSite(input: UpdateSiteInput!, id: String!): SiteUpdatedResponse!
  createPage(siteId: String!, input: CreatePageInput!): PageUpdatedResponse!
  deletePage(siteId: String!, id: String!): PageUpdatedResponse!
  updatePage(siteId: String!, id: String!, input: UpdatePageInput!): PageUpdatedResponse!
  getOrCreateFirstPage(siteId: String!): PageSchema!
  createComponent(input: CreateComponentInput!): ComponentSchema!
  updateComponent(id: String!, input: UpdateComponentInput!): Boolean!
  deleteComponent(id: String!): Boolean!
  archiveComponent(id: String!): Boolean!
  createElement(input: CreateElementInput!): String!
  updateElement(id: String!, input: UpdateElementInput!): Boolean!
  deleteElement(id: String!, input: DeleteElementInput!): Boolean!
}

input CreateSiteInput {
  """Site name"""
  name: String!
}

input UpdateSiteInput {
  """Site name"""
  name: String
}

input CreatePageInput {
  """Page title"""
  title: String!
}

input UpdatePageInput {
  """Page title"""
  title: String

  """Whether the title should be independent or part of the title template"""
  isLoneTitle: Boolean

  """Page description"""
  description: String
}

input CreateComponentInput {
  """Component's name"""
  name: String!
}

input UpdateComponentInput {
  """Component's name"""
  name: String
}

input CreateElementInput {
  """Site ID"""
  siteId: String!

  """Page ID"""
  pageId: String!

  """When is root, parent is not an element"""
  isRoot: Boolean

  """Parent of element"""
  parent: String!

  """Element type"""
  type: ElementType!

  """Additional properties"""
  props: JSON
}

input UpdateElementInput {
  """Parent of element"""
  parent: String

  """Element type"""
  type: ElementType

  """Additional properties"""
  props: JSON

  """Site ID"""
  siteId: String!

  """Page ID"""
  pageId: String!
}

input DeleteElementInput {
  """Site ID"""
  siteId: String!

  """Page ID"""
  pageId: String!
}

scalar link__Import