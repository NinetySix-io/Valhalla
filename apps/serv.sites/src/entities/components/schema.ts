import {
  BaseSchema,
  IsObjectId,
  SimpleModel,
  typegoose,
} from '@valhalla/serv.core';
import { Exclude, Expose } from 'class-transformer';
import { IsUrl, MaxLength, MinLength } from 'class-validator';

import { EditStatus } from '@app/protobuf';
import { ElementSchema } from '../elements/schema';
import { Field } from '@nestjs/graphql';
import mongoose from 'mongoose';

@SimpleModel('components')
@typegoose.index({ owners: 1 })
export class ComponentSchema extends BaseSchema {
  @typegoose.prop()
  @Expose()
  @Field({ description: 'Edit status' })
  status: EditStatus;

  @typegoose.prop()
  @Exclude()
  @Field({ description: 'Owners of the component' })
  owners: string[];

  @typegoose.prop()
  @Expose()
  @IsObjectId()
  @Field(() => String, { description: 'Account ID' })
  createdBy: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Expose()
  @IsObjectId()
  @Field(() => String, { description: 'Account ID' })
  updatedBy: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Expose()
  @MinLength(3)
  @MaxLength(30)
  @Field(() => String, { description: "Component's name" })
  name: string;

  @typegoose.prop()
  @Expose()
  @IsUrl()
  @Field(() => String, { description: 'Thumbnail URI', nullable: true })
  thumbnailUrl?: string;

  @typegoose.prop()
  @Exclude()
  elements: Array<ElementSchema>;
}
