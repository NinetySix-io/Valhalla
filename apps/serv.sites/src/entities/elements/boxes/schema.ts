import { Field, ObjectType } from '@nestjs/graphql';

import { ElementSchema } from '../schemas/element.schema';
import { Expose } from 'class-transformer';
import { HTMLType } from '@app/protobuf';
import { typegoose } from '@valhalla/serv.core';

@ObjectType()
export class BoxElementSchema extends ElementSchema {
  @typegoose.prop()
  @Expose()
  @Field(() => HTMLType)
  htmlType: HTMLType;
}
