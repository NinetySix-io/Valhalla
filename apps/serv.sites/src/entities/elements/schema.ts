import {
  BaseSchema,
  IsObjectId,
  SimpleModel,
  mongoose,
  typegoose,
} from '@valhalla/serv.core';
import { Exclude, Expose } from 'class-transformer';
import { Field, ObjectType, registerEnumType } from '@nestjs/graphql';

import { ElementType } from '@app/protobuf';
import { IsObject } from 'class-validator';

registerEnumType(ElementType, { name: 'ElementType' });

@ObjectType()
@SimpleModel('elements', { allowMixed: typegoose.Severity.ALLOW })
@typegoose.index({ parent: 1 })
export class ElementSchema extends BaseSchema {
  @typegoose.prop()
  @Expose()
  @Field({
    description: 'When is root, parent is not an element',
    nullable: true,
  })
  isRoot?: boolean;

  @typegoose.prop()
  @Exclude()
  ownBy: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Parent of element' })
  parent: string;

  @typegoose.prop()
  @Expose()
  @IsObjectId()
  @Field(() => String)
  updatedBy: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Element type' })
  type: ElementType;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Additional properties', nullable: true })
  @IsObject()
  props?: Record<string, unknown>;
}
