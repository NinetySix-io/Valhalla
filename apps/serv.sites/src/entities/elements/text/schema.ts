import { Field, ObjectType } from '@nestjs/graphql';

import { ElementSchema } from '../schemas/element.schema';
import { Expose } from 'class-transformer';
import { typegoose } from '@valhalla/serv.core';

@ObjectType()
export class TextElementSchema extends ElementSchema {
  @typegoose.prop()
  @Expose()
  @Field({ description: 'Text' })
  text: string;
}
