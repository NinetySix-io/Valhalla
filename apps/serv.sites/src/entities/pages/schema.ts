import {
  BaseSchema,
  SimpleModel,
  mongoose,
  typegoose,
} from '@valhalla/serv.core';
import { Exclude, Expose } from 'class-transformer';
import { Field, ObjectType, registerEnumType } from '@nestjs/graphql';

import { PageStatus } from '@app/protobuf';

registerEnumType(PageStatus, {
  name: 'PageStatus',
});

@ObjectType()
@SimpleModel('pages')
@typegoose.index({ site: 1 })
@typegoose.index({ ownBy: 1 })
@typegoose.index(
  { ownBy: 1, site: 1, slug: 1 },
  {
    unique: true,
    partialFilterExpression: {
      slug: { $type: 'string' },
    },
  },
)
export class PageSchema extends BaseSchema {
  @typegoose.prop()
  @Exclude()
  ownBy: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Exclude()
  site: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Page url', nullable: true })
  slug?: string;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Page deployment status' })
  status: PageStatus;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Page title' })
  title: string;

  @typegoose.prop()
  @Expose()
  @Field({
    nullable: true,
    description:
      'Whether the title should be independent or part of the title template',
  })
  isLoneTitle?: boolean;

  @typegoose.prop()
  @Expose()
  @Field({ description: 'Page description', nullable: true })
  description?: string;

  @typegoose.prop()
  @Exclude()
  @Field(() => String, { description: 'Account ID of creator' })
  createdBy: mongoose.Types.ObjectId;

  @typegoose.prop()
  @Exclude()
  @Field(() => String, { description: 'Account ID of updater' })
  updatedBy: mongoose.Types.ObjectId;
}
