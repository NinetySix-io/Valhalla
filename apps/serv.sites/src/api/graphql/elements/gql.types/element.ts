import { Expose, Transform, Type } from 'class-transformer';
import { Field, ObjectType, registerEnumType } from '@nestjs/graphql';
import { PageElement, PrimitiveElementType } from '@app/protobuf';

import { ElementArea } from './element.area';
import { IsObjectId } from '@valhalla/serv.core';
import { ObjectIDResolver } from 'graphql-scalars';
import { PageElementProto } from '@app/cqrs/protos/page.element.proto';
import { ValidateNested } from 'class-validator';

registerEnumType(PrimitiveElementType, { name: 'ElementType' });

@ObjectType({ isAbstract: true })
export class Element implements Omit<PageElementProto, 'type'> {
  @Expose()
  @Field(() => ObjectIDResolver)
  @IsObjectId()
  id: string;

  @Expose()
  @Field(() => PrimitiveElementType)
  @Transform((p: { obj: PageElement }) =>
    p.obj.type.$case === 'text' ? PrimitiveElementType.TEXT : null,
  )
  type: PrimitiveElementType;

  @Expose()
  @Field(() => ObjectIDResolver)
  @IsObjectId()
  updatedBy: string;

  @Expose()
  @Field(() => ObjectIDResolver)
  @IsObjectId()
  createdBy: string;

  @Expose()
  @Field(() => ObjectIDResolver)
  @IsObjectId()
  group: string;

  @Expose()
  @Field(() => ElementArea)
  @Type()
  @ValidateNested()
  desktop: ElementArea;

  @Expose()
  @Field(() => ElementArea, { nullable: true })
  @Type()
  @ValidateNested()
  tablet?: ElementArea;

  @Expose()
  @Field(() => ElementArea, { nullable: true })
  @Type()
  @ValidateNested()
  mobile?: ElementArea;

  @Expose()
  @Field({ nullable: true })
  @ValidateNested()
  createdAt?: Date;

  @Expose()
  @Field({ nullable: true })
  updatedAt?: Date;
}
