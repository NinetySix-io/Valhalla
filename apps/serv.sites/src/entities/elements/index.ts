import { Injectable } from '@nestjs/common';
import { InjectModel } from 'nestjs-typegoose';
import { BaseFactory, ModelType } from '@valhalla/serv.core';
import { ElementSchema } from './schema';

@Injectable()
export class ElementsModel extends BaseFactory<ElementSchema> {
  constructor(
    @InjectModel(ElementSchema)
    model: ModelType<ElementSchema>,
  ) {
    super(model);
  }
}
