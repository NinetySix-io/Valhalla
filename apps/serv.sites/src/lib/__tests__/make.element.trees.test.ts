import { ElementSchema } from '@app/entities/elements/schema';
import { ElementType } from '@app/protobuf';
import { makeElementTrees } from '../make.element.trees';
import { toObjectId } from '@valhalla/serv.core';

describe('Lib:: makeElementTrees', () => {
  const owner = toObjectId();
  const leafEnd = toObjectId('62cba8b78eb6823f125696a6');
  const leaf1 = toObjectId('62cba8b78eb6823f125696a3');
  const parent1 = toObjectId('62cba8b78eb6823f125696a4');
  const parent2 = toObjectId('62cba8b78eb6823f125696a5');

  function createSeed(
    prop: Pick<ElementSchema, '_id' | 'isRoot' | 'type' | 'parent'>,
  ): ElementSchema {
    return {
      id: prop._id.toHexString(),
      _id: prop._id,
      isRoot: prop.isRoot,
      type: prop.type,
      parent: prop.parent,
      ownBy: owner,
      createdAt: new Date(),
      updatedAt: new Date(),
      updatedBy: owner,
    };
  }

  const goodStruct: ElementSchema[] = [
    createSeed({
      _id: parent1,
      isRoot: true,
      type: ElementType.a,
      parent: '',
    }),
    createSeed({
      _id: leaf1,
      isRoot: false,
      type: ElementType.a,
      parent: parent1.toHexString(),
    }),
    createSeed({
      _id: leafEnd,
      isRoot: false,
      type: ElementType.a,
      parent: leaf1.toHexString(),
    }),
    createSeed({
      _id: parent2,
      isRoot: true,
      type: ElementType.a,
      parent: '',
    }),
  ];

  it('make a good tree', () => {
    const tree = makeElementTrees(goodStruct);
    expect(tree).toMatchSnapshot();
  });
});
