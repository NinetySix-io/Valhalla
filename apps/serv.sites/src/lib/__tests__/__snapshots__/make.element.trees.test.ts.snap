// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lib:: makeElementTrees make a good tree 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": "62cba8b78eb6823f125696a6",
            "props": undefined,
            "type": "a",
          },
        ],
        "id": "62cba8b78eb6823f125696a3",
        "props": undefined,
        "type": "a",
      },
    ],
    "id": "62cba8b78eb6823f125696a4",
    "props": undefined,
    "type": "a",
  },
  Object {
    "children": Array [],
    "id": "62cba8b78eb6823f125696a5",
    "props": undefined,
    "type": "a",
  },
]
`;
